src/nix/comment.cc

```diff
@@ -0,0 +1,227 @@
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <regex>
+#include <climits>
+#include <algorithm>
+
+#include "comment.hh"
+#include "util.hh"
+
+// This module looks for documentation comments in the source code.
+//
+// Documentation is not retained during parsing, and it should not be,
+// for performance reasons. Because of this the code has to jump
+// through some hoops, to perform its task.
+//
+// Adapting the parser was not considered an option, so this code
+// parses the comments from scratch, using regular expressions. These
+// do not support all syntactic constructs, so in rare cases, they
+// will fail and the code will report no documentation.
+//
+// One such situation is where documentation is requested for a
+// partially applied function, where the outer lambda pattern
+// matches an attribute set. This is not supported in the regexes
+// because it potentially requires (almost?) the entire grammar.
+//
+// This module has been designed not to report the wrong
+// documentation; considering that the wrong documentation is worse
+// than no documentation. The regular expressions will only match
+// simple, well understood syntactic structures, or not match at all.
+//
+// This approach to finding documentation does not cause extra runtime
+// overhead, until used.
+//
+// This module does not support tab ('\t') characters. In some places
+// they are treated as single spaces. They should be avoided.
+
+namespace nix::Comment {
+
+struct Doc emptyDoc("", "", "", 0);
+
+// parseDoc will try to recover a Doc by looking at the text that leads up to a term
+// definition.
+static struct Doc parseDoc(std::string sourcePrefix);
+
+// stripComment unpacks a comment, by unindenting and stripping " * " prefixes as
+// applicable. The argument should include any preceding whitespace.
+static std::string stripComment(std::string rawComment);
+
+// Consistent unindenting. It will only remove entire columns.
+static std::string unindent(std::string s);
+
+static std::string trimUnindent(std::string s) {
+    return trim(unindent(s));
+}
+
+static std::string stripPrefix(std::string prefix, std::string s) {
+  std::string::size_type index = s.find(prefix);
+  return (index == 0) ? s.erase(0, prefix.length()) : s;
+}
+
+struct Doc lookupDoc(Pos & pos) {
+
+    try {
+        std::ifstream ifs(static_cast<const std::string>(pos.file));
+        std::stringstream buffer;
+
+        // read up to line and column with unix newlines
+
+        std::string line;
+        for (size_t i = 0;
+             getline(ifs, line) && i < pos.line-1;
+             i++) {
+            buffer << line << "\n";
+        }
+        buffer << line.substr(0, pos.column-1);
```

Comment:
This can be simplified to something like `for (auto & line : tokenizeString(readFile(pos.file), "\n") { ... }`.

Reply:
`tokenizeString` didn't work because it doesn't give back the empty lines. I didn't find another suitable function in util either, so I have factored the thing out instead. The other review items are now solved.